src/header.rs

```diff
@@ -82,6 +88,32 @@ pub fn add_header(headers: &mut Vec<Header>, header: Header) {
     headers.push(header);
 }
 
+// https://tools.ietf.org/html/rfc7230#section-3.2.3
+// Each header field consists of a case-insensitive field name followed
+// by a colon (":"), optional leading whitespace, the field value, and
+// optional trailing whitespace.
+// field-name     = token
+// token = 1*tchar
+// tchar = "!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" / "." /
+// "^" / "_" / "`" / "|" / "~" / DIGIT / ALPHA
+fn valid_name(name: &str) -> bool {
+    name.len() > 0 && name.bytes().all(is_tchar)
+}
+
+#[inline]
+fn is_tchar(b: u8) -> bool {
+    match b {
+        b'!' | b'#' | b'$' | b'%' | b'&' | b'\'' | b'*' | b'+' | b'-' | b'.' | b'^' | b'_'
+        | b'`' | b'|' | b'~' | b'A' | b'B' | b'C' | b'D' | b'E' | b'F' | b'G' | b'H' | b'I'
+        | b'J' | b'K' | b'L' | b'M' | b'N' | b'O' | b'P' | b'Q' | b'R' | b'S' | b'T' | b'U'
+        | b'V' | b'W' | b'X' | b'Y' | b'Z' | b'a' | b'b' | b'c' | b'd' | b'e' | b'f' | b'g'
+        | b'h' | b'i' | b'j' | b'k' | b'l' | b'm' | b'n' | b'o' | b'p' | b'q' | b'r' | b's'
+        | b't' | b'u' | b'v' | b'w' | b'x' | b'y' | b'z' | b'0' | b'1' | b'2' | b'3' | b'4'
+        | b'5' | b'6' | b'7' | b'8' | b'9' => true,
```

Comment:
Maybe you prefer this for clarity (or speed?), but just pointing out that there are these in the std lib:

ALPHA => https://doc.rust-lang.org/beta/std/primitive.u8.html#method.is_ascii_alphabetic
DIGIT => https://doc.rust-lang.org/beta/std/primitive.u8.html#method.is_ascii_digit

or even the combo:

https://doc.rust-lang.org/beta/std/primitive.u8.html#method.is_ascii_alphanumeric

Reply:
Ah, thanks for the reminder! I will admit, there was part of me that figured this might be a little speedier, but it's premature optimization; I prefer to use the clearer function and possibly optimize later if it turns out to be a bottleneck.